packages/web-forms/tests/components/OdkWebForm.test.ts

```diff
@@ -0,0 +1,64 @@
+import OdkWebForm from '@/components/OdkWebForm.vue';
+import { flushPromises, mount } from '@vue/test-utils';
+import { describe, expect, it } from 'vitest';
+import { getFormXml, globalMountOptions } from '../helpers';
+const mountComponent = () => {
+	const xform = getFormXml('validation/2-simple-required.xml');
+
+	const component = mount(OdkWebForm, {
+		props: {
+			formXml: xform,
+		},
+		global: globalMountOptions,
+		attachTo: document.body,
+	});
+
+	return component;
+};
+
+describe('OdkWebForm', () => {
+	if (!Element.prototype.scrollIntoView) {
+		// eslint-disable-next-line @typescript-eslint/no-empty-function
+		Element.prototype.scrollIntoView = () => {};
+	}
+	if (!HTMLElement.prototype.showPopover) {
+		HTMLElement.prototype.showPopover = function () {
+			this.style.display = 'block';
+		};
+	}
+	if (!HTMLElement.prototype.hidePopover) {
+		HTMLElement.prototype.hidePopover = function () {
+			this.style.display = 'none';
+		};
+	}
```

Comment:
I know there's a lot here, but it's worth considering. There are a few good candidates for moving to somewhere shared/reusable, but we can start here for now.

Main goals:

1. Clean up any mutation of global scope/prototypes
2. Use mocking facilities consistently
3. Generalize enough that we can apply similar techniques without a lot of extra fuss in the future

```suggestion
	// TODO: how the heck is `undefined` a key of anything?!
	type StringKeyOf<T> = Extract<keyof T, string>;

	type StringKeyOfDocument = StringKeyOf<Document>;

	// prettier-ignore
	type DocumentPropertyName = {
		[PropertyName in StringKeyOfDocument]:
			Document[PropertyName] extends AnyFunction
				? never
				: PropertyName;
	}[StringKeyOfDocument];

	let documentKeysAdded: StringKeyOfDocument[] = [];

	const mockDocumentGetter = <PropertyName extends DocumentPropertyName>(
		propertyName: PropertyName,
		mockImplementation: () => Document[PropertyName]
	) => {
		if (propertyName in document) {
			return vi.spyOn(document, propertyName, 'get').mockImplementation(mockImplementation);
		}

		documentKeysAdded.push(propertyName);

		const mock = vi.fn(mockImplementation);

		Object.defineProperty(document, propertyName, {
			get: mock,
			configurable: true,
		});

		return mock;
	};

	type StringKeyOfHTMLElement = StringKeyOf<HTMLElement>;

	// prettier-ignore
	type ElementMethodName = {
		[PropertyName in StringKeyOfHTMLElement]:
			HTMLElement[PropertyName] extends AnyFunction
				? PropertyName
				: never;
	}[keyof HTMLElement];

	type ElementMethodMock<MethodName extends ElementMethodName> = (
		this: HTMLElement,
		...args: Parameters<HTMLElement[MethodName]>
	) => ReturnType<HTMLElement[MethodName]>;

	let elementKeysAdded: StringKeyOfHTMLElement[] = [];

	const mockElementPrototypeMethod = <MethodName extends ElementMethodName>(
		methodName: MethodName,
		mockImplementation: ElementMethodMock<MethodName>
	) => {
		if (methodName in HTMLElement.prototype) {
			return vi.spyOn(HTMLElement.prototype, methodName).mockImplementation(function (...args) {
				// eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
				return mockImplementation.apply(this as HTMLElement, args as any);
			});
		}

		elementKeysAdded.push(methodName);

		const mock = vi.fn(mockImplementation);

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
		HTMLElement.prototype[methodName] = mock as any;

		return mock;
	};

	beforeEach(() => {
		documentKeysAdded = [];
		elementKeysAdded = [];

		// Not necessary if we scroll `document.scrollingElement` instead.
		// Including to show how it'd be done if not.
		vi.stubGlobal('scrollTo', () => {
			/* Do nothing, unless we test this behavior. */
		});

		mockDocumentGetter('scrollingElement', () => {
			return document.documentElement;
		});

		mockElementPrototypeMethod('showPopover', function () {
			this.style.display = 'block';
		});

		mockElementPrototypeMethod('hidePopover', function () {
			this.style.display = 'none';
		});
	});

	afterEach(() => {
		documentKeysAdded.forEach((propertyName) => {
			delete document[propertyName];
		});

		elementKeysAdded.forEach((methodName) => {
			delete HTMLElement.prototype[methodName];
		});

		vi.restoreAllMocks();
		vi.unstubAllGlobals();
	});
```

(You'll also need these import changes:)

```diff
- import { describe, expect, it } from 'vitest';
+ import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
+ import type { AnyFunction } from '@getodk/common/types/helpers.d.ts';
```